pulser/simulation/simresults.py

```diff
@@ -14,63 +14,336 @@
+
+    @property
+    def results(self) -> list[Counter]:
+        """Probability distribution of the bitstrings."""
+        return self._results
+
+    def get_state(self, t: float, t_tol: float = 1.e-3) -> qutip.Qobj:
+        """Get the state at time t as a diagonal density matrix.
+
+        Note:
+            This is not the density matrix of the system, but is a convenient
+            way of computing expectation values of observables.
+
+        Args:
+            t (int): Time of the state to be returned.
+            t_tol (float): Tolerance for the difference between t and
+                closest time.
+
+        Returns:
+            qutip.Qobj: States probability distribution as a diagonal
+                density matrix.
+        """
+        def _proj_from_bitstring(bitstring: str) -> qutip.Qobj:
+            # In the digital case, |h> = |1> = qutip.basis()
+            if self._basis_name == 'digital':
+                proj = qutip.tensor([qutip.basis(2, int(i)).proj() for i
+                                     in bitstring])
+            # ground-rydberg basis case
+            else:
+                proj = qutip.tensor([qutip.basis(2, 1-int(i)).proj() for i
+                                     in bitstring])
+            return proj
+
+        t_index = self._get_index_from_time(t, t_tol)
+        return sum(v * _proj_from_bitstring(b) for
+                   b, v in self._results[t_index].items())
+
+    def get_final_state(self) -> qutip.Qobj:
+        """Get the final state of the simulation as a diagonal density matrix.
+
+        Note: This is not the density matrix of the system, but is a convenient
+            way of computing expectation values of observables.
+
+        Returns:
+            qutip.Qobj: States probability distribution as a density matrix.
+        """
+        return self.get_state(self._sim_times[-1])
+
+    def expect(self, obs_list: Sequence[Union[qutip.Qobj, ArrayLike]]
+               ) -> list[Union[float, complex, ArrayLike]]:
+        """Calculates the expectation value of a list of observables.
+
+        Args:
+            obs_list (array-like of qutip.Qobj or array-like of numpy.ndarray):
+                A list of observables whose expectation value will be
+                calculated. If necessary, each member will be transformed into
+                a qutip.Qobj instance.
+
+        Note: This only works for diagonal observables, since results have been
+            projected onto the Z basis.
+
+        Returns:
+            list: List of expectation values of each operator.
+        """
+        for obs in obs_list:
+            if not isdiagonal(obs):
+                raise ValueError(f"Observable {obs} is non-diagonal.")
+
+        return super().expect(obs_list)
+
+    def _calc_weights(self, t_index: int) -> list[float]:
+        N = self._size
+        bitstrings = [np.binary_repr(k, N) for k in range(2**N)]
+        return [self._results[t_index][b] for b in bitstrings]
```

Comment:
Are you going through all the bitstrings in `self._results` here?

Reply:
Yes, we need to have probabilities for each possible bitstring in order to sample afterwards. I'm sure there is a more efficient way, but I haven't much thought about it....

Reply:
Would this be equivalent?
```suggestion
        return list(self._results[t_index].values())
```
Assuming the values are in the right order, I guess.

Reply:
No, sadly... When you sample from this `weights`list, you need to have all bitstrings inside, even those with zero probability, or else you wouldn't get the bitstrings in the right order afterwards. For example if `weights = [0., 1.]`in a Hilbert space of dimension `4`, assuming that `1.`means 'probability 1 for bit string 11', you wouldn't get bitstring 11 after the multinomial sampling since you don't have enough weights. I'm still sure there's a better method though.

Reply:
Like, if you could sample from a `dict`, not from a `list`, so you could specify the bitstrings you want.

Reply:
But for a given `self._results[t_index]` counter, doesn't it have an entry for all possible bitstrings?

Reply:
No, we chose not to include zero-probability docstrings. Maybe that was a design flaw ? Do you think we should include them ?

Reply:
No, I think this way is good, it's a good use of the `Counter` class. In that case, I suggest this (smaller) improvement:
```suggestion
        return [self._results[t_index][np.binary_repr(k, N)] for k in range(2**N)]
```

Reply:
If you feel like having a bunch of `0` weights on the multinomial distirbution is costing you, there is a way around it... But it's not super clean, so I'm not sure it's worth it

Reply:
Alright. Anyway, we don't have to call this method much, contrary to `CleanResults.sample_state`, so it should be no big deal to have `O(2**n)`complexity just this once.